test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,202 @@
+		if (subNode != _property)
+			return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+		if (_root.isMember(subNode))
+			return make_optional(_root[subNode]);
+	}
+	return {};
+}
+
+void ASTPropertyTest::readTestedProperties(Json::Value const& _astJson)
+{
+	queue<Json::Value> nodesToVisit;
+	nodesToVisit.push(_astJson);
+
+	set<string> readTests;
+
+	while (!nodesToVisit.empty())
+	{
+		Json::Value& node = nodesToVisit.front();
+
+		if (node.isArray())
+			for (auto&& member: node)
+				nodesToVisit.push(member);
+		else if (node.isObject())
+			for (string const& memberName: node.getMemberNames())
+			{
+				if (memberName != "documentation")
+				{
+					nodesToVisit.push(node[memberName]);
+					continue;
+				}
+
+				string nodeDocstring = node["documentation"].isObject() ?
+					node["documentation"]["text"].asString() :
+					node["documentation"].asString();
+				soltestAssert(!nodeDocstring.empty());
+
+				vector<StringPair> pairs = readKeyValuePairs(nodeDocstring);
+				soltestAssert(!pairs.empty(), "It is required at least one property test per annotated AST node.");
+
+				for (auto const& [testId, testedProperty]: pairs)
+				{
+					soltestAssert(
+						readTests.insert(testId).second,
+						"Test \"" + testId + "\" was already defined before."
+					);
+					soltestAssert(
+						m_testCases.count(testId) > 0,
+						"Test \"" + testId + "\" does not have a corresponding expected value."
+					);
+					m_testCases[testId].property = testedProperty;
+
+					soltestAssert(node.isMember("nodeType"));
+					optional<Json::Value> propertyNode = findNode(node, testedProperty);
+					soltestAssert(
+						propertyNode.has_value(),
+						node["nodeType"].asString() + " node does not have a property named \""s + testedProperty + "\""
+					);
+					soltestAssert(
+						!propertyNode->isObject() && !propertyNode->isArray(),
+						"Property \"" + testedProperty + "\" is an object or an array."
+					);
+					m_testCases[testId].obtainedValue = propertyNode->asString();
+				}
+			}
+
+		nodesToVisit.pop();
+	}
+
+	auto missingPropertyTest = ranges::find_if(
+		m_testCases,
+		[&](auto const& _testCase) { return readTests.count(_testCase.first) == 0; }
+	);
+	soltestAssert(
+		missingPropertyTest == ranges::end(m_testCases),
+		"Missing annotated property for test \"" + missingPropertyTest->first + "\""
+	);
+
+	generateTestCaseValues(m_obtainedResult, true);
+}
+
+TestCase::TestResult ASTPropertyTest::run(ostream& _stream, string const& _linePrefix, bool const _formatted)
+{
+	CompilerStack compiler;
+
+	compiler.setSources({{
+		"A",
+		"pragma solidity >=0.0;\n// SPDX-License-Identifier: GPL-3.0\n" + m_source
+	}});
+	compiler.setEVMVersion(solidity::test::CommonOptions::get().evmVersion());
+	compiler.setOptimiserSettings(solidity::test::CommonOptions::get().optimize);
+	if (!compiler.parseAndAnalyze())
+		BOOST_THROW_EXCEPTION(runtime_error("Parsing contract failed"));
```

Comment:
If you do this, the only thing you'll see when the test has compilation errors will be `Parsing contract failed`, without any details about where the error is.

Please do at least this:
```suggestion
		BOOST_THROW_EXCEPTION(runtime_error(
			"Parsing contract failed:\n" +
			SourceReferenceFormatter::formatErrorInformation(compiler.errors(), compiler, _formatted)
		));
```

Ideally, your `TestCase` would inherit from `AnalysisFramework` instead and use its compiler stack and error formatting functions. Though for now just adding errors to the message will be fine. We can refactor this later.